### PR TITLE
Pin Rust version in Dockerfile

### LINKED ISSUE #58: Update Rust version
Hi, so I've used a relatively new Rust feature in one of my exercises. ([scoped threads](https://doc.rust-lang.org/std/thread/fn.scope.html), introduced in 1.63, about two months ago)

My tests fail and it's saying that scoped threads are an unstable library version. I'm assuming the Rust version of the test runner is lower than 1.63.

Can this version be bumped? (I will gladly help out if it makes sense, but I'm not very experienced with docker.)

Thank you for taking a look!

### PR DESCRIPTION
**Pull request by remlse**
closes #58

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # always build this using the latest stable release
-FROM rust:latest as build
+FROM rust:1.64.0 as build
```
[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply Dockerfile:2]: A bit later in this file goes the line
```
FROM rustlang/rust:nightly AS test
```

That is what is used to actually run the tests.
When this runner gets rebuilt, there are good chances that some ancient version of `rustlang/rust:nightly` pops out of the cache.
Unfortunately, the maintainers of `rustlang/rust` do not seem to have provided means to point to some exact version of nightly.

I would work it around like that:

```
FROM rust:1.64 as test
RUN rustup toolchain add nightly
```

**Review by RGafiyatullin — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply Dockerfile:2]: (@remlse I am not sure how your notifications are set up at Github.)

Would you please have a look above?

File: Dockerfile [hunk 2]
```diff
@@ -37,7 +37,8 @@ COPY local-registry/* ./
 RUN ${wd}/bin/generate-registry.sh
 
 # As of Dec 2019, we need to use the nightly toolchain to get JSON test output
-FROM rustlang/rust:nightly AS test
+FROM rust/nightly AS test
```
Comment (RGafiyatullin) [inline Dockerfile:40]: `rust/nightly` has the same problem with caching as `rustlang/rust:nightly` does.

The idea is to use some "absolute" (as opposed to "relative") tag as a base image.
Please consider using `FROM rust:1.64`: it does not have to be a "nightly", we are adding the nightly toolchain with the very next line.

File: Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (remlse) [reply Dockerfile]: Thanks for taking a look!

I made the changes. I tried to build the docker image locally, but it didn't work. But I'm getting the same error on the main branch, so I think it's unrelated.

```
error: failed to sync

Caused by:
  failed to load pkg lockfile

Caused by:
  failed to select a version for the requirement `bstr = "=1.0.1"`
  candidate versions found which didn't match: 0.2.17, 0.2.16, 0.2.15, ...
  location searched: crates.io index
  required by package `dummy_package v0.1.0 (/local-registry)`
The command '/bin/sh -c ${wd}/bin/generate-registry.sh' returned a non-zero code: 101
```

File: Dockerfile (hunk 2 above)
Comment (RGafiyatullin) [reply Dockerfile]: I would also not include the `Cargo.lock`. Why changing two aspects of the someone else's system with a single PR? :)

Comment (remlse) [reply Dockerfile]: Sorry about that. Must've happened automatically, I didn't `cargo update` or anything

File: Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply Dockerfile]: Indeed it looks like some strange stuff with cargo-local-registry :\

**Review by RGafiyatullin — APPROVED** [review]
LGTM.

---
##### about the image build failure

The image build-failure probably was caused by some transient issues of the combination of crates.io + cargo + cargo-local-registry.

Right now I cannot reproduce the build-failure.

**Review by ErikSchierboom — COMMENTED** [review]
I would also pin the version used at the top: https://github.com/exercism/rust-test-runner/blob/0678caa9558134e4b0b27695c4f7c3c16c7005e0/Dockerfile#L2

**Comment by RGafiyatullin** [comment]
@ErikSchierboom , when is that expected to be deployed?

**Comment by ErikSchierboom** [comment]
That happens automatically, so within about 10 minutes of when this was merged.

**Comment by remlse** [comment]
Hm, I tried to resubmit "parallel letter frequency" where I used scoped threads, and I still got the same error. In case the test output is relevant, [my solution is here](https://exercism.org/tracks/rust/exercises/parallel-letter-frequency/solutions/remlse).

What could be the problem?

**Comment by remlse** [comment]
looks like CI is failing, sadly I don't know what's going on looking at the output.

**Comment by remlse** [comment]
@ErikSchierboom I've seen that, but this feature has been stabilized in Rust 1.64. So if the Rust version was the recent one, this error would not occur. I believe the test runner used is the same as before - with the outdated Rust version - because CI and the deployment in failing. (for reasons I don't understand)

**Comment by ErikSchierboom** [comment]
Ah, I saw this in the deployment log:

> "/bin/sh -c ${wd}/bin/generate-registry.sh" did not complete successfully

The more extensive output is:

```shell
#23 57.89 error: failed to sync
#23 57.89 
#23 57.89 Caused by:
#23 57.89   failed to load pkg lockfile
#23 57.89 
#23 57.89 Caused by:
#23 57.89   failed to select a version for the requirement `bstr = "=1.0.1"`
#23 57.89   candidate versions found which didn't match: 0.2.17, 0.2.16, 0.2.15, ...
#23 57.89   location searched: crates.io index
#23 57.89   required by package `dummy_package v0.1.0 (/local-registry)`
#23 ERROR: process "/bin/sh -c ${wd}/bin/generate-registry.sh" did not complete successfully: exit code: 101
------
 > [build 15/15] RUN /rust-test-runner/bin/generate-registry.sh:
: failed to sync
#23 57.89 
#23 57.89 Caused by:
#23 57.89   failed to load pkg lockfile
#23 57.89 
#23 57.89 Caused by:
#23 57.89   failed to select a version for the requirement `bstr = "=1.0.1"`
#23 57.89   candidate versions found which didn't match: 0.2.17, 0.2.16, 0.2.15, ...
#23 57.89   location searched: crates.io index
#23 57.89   required by package `dummy_package v0.1.0 (/local-registry)`
```

Any idea?

**Comment by remlse** [comment]
Sadly, no. This seems really weird, version `1.0.1` of `bstr` is definitely available on crates.io. `cargo add bstr` defaults to version `1.0.1` too.

https://crates.io/crates/bstr/versions

I would have to get into how this whole test runner works, which I don't know if and when I might have time to do.

**Comment by ErikSchierboom** [comment]
@dhovart Do you have any idea how to resolve this issue?

**Pull request merged**